### DISCUSSION TITLE
RestAPI: permission denied for schema extensions

### DISCUSSION THREAD

**Discussion by ghost** [supabase/supabase#2462] [Questions] [answered]
By default the PostgREST json-format for adding and updating records is differently, **except** when you add records with an prefilled primary key. 

So I created the function

```
CREATE FUNCTION generate_uuid()
RETURNS text AS $$
 SELECT extensions.uuid_generate_v4();
$$ LANGUAGE SQL IMMUTABLE;
```

This works fine locally, but NOT via PostgREST via POST /rpc/generate_uuid HTTP/1.1 method.

How to solve this? How can I give permission for extensions.uuid_generate_v4 ?

**Comment by steve-chavez** [comment, marked as answer]
This should do:

```sql
grant usage on schema extensions to anon, authenticated, service_role;
```

Strange though, this privilege is added by default. Did you dropped and recreated the schema extensions by any chance?

**Reply by ghost** [reply]
This solved it! I however didn't touch the schema extensions at all. As https://postgrest.org/en/stable/api.html#stored-procedures however stated "Whenever you create or change a function you must refresh PostgREST’s schema cache." I restarted to server hoping this would trigger a refresh....

Great, thanks!

**Reply by steve-chavez** [reply]
@ptr008 :O there shouldn't be a need to do that because PostgREST is refreshed automatically in Supabase via an event trigger(which will also be added to postgrest-docs soon). In case you somehow dropped it, check this: https://github.com/supabase/supabase/issues/2200#issuecomment-873503353
